### PR TITLE
GG-34170 C++ thin: SSL certificates, keys and CA are not mandatory

### PR DESCRIPTION
**Pull request by isapego**
- Improved error reporting during SSL failures.
- Fixed issue when client compiled and run with different OpenSSL versions.
- Re-factored to remove SSL-related code duplication.
- Added missing tests to some test suites.
- Certificate, private key and CA are not mandatory parameters for client now. Also, implemented extracting and usage of system CA for Windows by default.

### CODE REVIEW

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/platforms/cpp/network/src/network/ssl/secure_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    {
+        using namespace ignite::network::ssl;
+
+        assert(sslContext != 0);
+        SslGateway &sslGateway = SslGateway::GetInstance();
+
+#ifndef _WIN32
+        long res = sslGateway.SSL_CTX_set_default_verify_paths_(sslContext);
+        if (res != SSL_OPERATION_SUCCESS)
+            ThrowSecureError("Can not set default Certificate Authority for secure connection. Try setting custom CA");
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/network/src/network/ssl/secure_utils.cpp:42]: The problem is, the same library is used for ODBC and thin client and API is different for them.

File: modules/platforms/cpp/network/src/network/ssl/secure_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+                sslGateway.SSL_CTX_set_verify_depth_(sslContext, 8);
+
+                sslGateway.SSL_CTX_set_options_(sslContext, SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3 | SSL_OP_NO_COMPRESSION);
+
+                if (!cfg.caPath.empty())
+                {
+                    long res = sslGateway.SSL_CTX_load_verify_locations_(sslContext, cfg.caPath.c_str(), 0);
+                    if (res != SSL_OPERATION_SUCCESS)
+                        ThrowSecureError("Can not set Certificate Authority path for secure connection");
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/network/src/network/ssl/secure_utils.cpp:105]: Will do

File: modules/platforms/cpp/network/src/network/ssl/secure_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        SslGateway &sslGateway = SslGateway::GetInstance();
+
+#ifndef _WIN32
+        long res = sslGateway.SSL_CTX_set_default_verify_paths_(sslContext);
+        if (res != SSL_OPERATION_SUCCESS)
+            ThrowSecureError("Can not set default Certificate Authority for secure connection. Try setting custom CA");
+#else
+        X509_STORE *sslStore = sslGateway.X509_STORE_new_();
+        if (!sslStore)
+            ThrowSecureError("Can not create X509_STORE certificate store. Try setting custom CA");
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/network/src/network/ssl/secure_utils.cpp:46]: Not sure how to handle error here properly. Documentation says nothing about diagnostics and there is a high chance there will be "No error" if we just look into error stack.

**Review by ptupitsyn — APPROVED** [review]
Looks good to me.

Can we add any tests for the new error handling code?

**Pull request merged**